### PR TITLE
Add docs for cloning feature

### PR DESCRIPTION
**Pull request by j-griffith**
Doc additions describing the clone feature and how to implement.

### CODE REVIEW

File: book/src/external-provisioner.md [hunk 1]
```diff
@@ -25,6 +25,8 @@ The deletion of a `PersistentVolumeClaim` object bound to a `PersistentVolume` c
 
 The CSI `external-provisioner` also supports the `Snapshot` DataSource. If a `Snapshot` CRD is specified as a data source on a PVC object, the sidecar container fetches the information about the snapshot by fetching the `SnapshotContent` object and populates the data source field in the resulting `CreateVolume` call to indicate to the storage system that the new volume should be populated using the specified snapshot.
 
+Cloning is also implemented by specifying a `kind:` of type `PersistentVolumeClaim` in the DataSource field of a Provision request.  It's the responsbility of the external-provisioner to verify that the claim specified in the DataSource object exists, is in the same storage class as the volume being provisioned and that the claim is currently `Bound`.
```
Comment (msau42) [inline book/src/external-provisioner.md]: Would it be useful to have a subsection just for DataSources?

File: book/src/volume-cloning.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
+# Volume Cloning
+
+## Status
+
+* Kubernetes 1.15: Alpha
```
Comment (msau42) [inline book/src/volume-cloning.md]: Can you put this in table format similar to the other feature pages?

File: book/src/volume-cloning.md [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+## Overview
+
```
Comment (msau42) [inline book/src/volume-cloning.md]: Maybe start off with one sentence about what cloning is, and then a second sentence pointing to the Kubernetes user docs.

File: book/src/volume-cloning.md [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+A storage provider that allows volume cloning as a create feature, may choose to implement volume cloning via a
+control-plane CSI RPC call.
+
+## Implementing Volume cloning functionality
+
+To implement volume cloning the CSI driver MUST:
+
+1. Implement checks in `CreateVolume` function implementation.
```
Comment (msau42) [inline book/src/volume-cloning.md]: What parameters are set in CreateVolume that indicate this should be a clone?  This should be detailed in this section

File: book/src/volume-cloning.md [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+2. Implement `CLONE_VOLUME` controller capability.
+
+The cloning feature in Kubernetes is implemented by adding a `DataSource` entry of `kind: PersistentVolumeClaim` to the csi.CreateVolumeRequest.  This presents the source volume (PersistentVolumeClaim to clone from) as a generic data source.  The source volume must:
```
Comment (msau42) [inline book/src/volume-cloning.md]: Unsure if we need to document this part about Kubernetes user requirements. There's nothing a driver author can do for this right?

File: book/src/volume-cloning.md [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+It is the responsibility of the storage plugin to either implement an expansion after clone if a provision request size is greater than the source.
+
+## Deploying volume clone functionality
+
+The Kubernetes CSI development team maintains the [external-provisioner](external-provisioner.md) which is responsible for detecting requests for a PVC DataSource and providing that information to the plugin during provisionion in the call to the plugins `CreateVolume` function.  It's up to the plugin to check the `csi.CreateVolumeRequest` for a `VolumeContentSource` entry in the CreateVolumeRequest object.
+
+There are no additional side-cars or add on components required.
+
+## Enabling Volume expansion for CSI volumes in Kubernetes
```
Comment (msau42) [inline book/src/volume-cloning.md]: Title needs update

File: book/src/volume-cloning.md [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+Volume cloning for CSI volumes is an alpha feature (Kubernetes 1.15) and hence must be explicitly enabled via feature gate:
+
+```
+--feature-gates=VolumePVCDataSource=true
+```
+
+## Example implementation
+
+A trivial example implementation can be found in the [csi-hostpath plugin](https://github.com/kubernetes-csi/csi-driver-host-path) in it's implementation of `CreateVolume`.
```
Comment (msau42) [inline book/src/volume-cloning.md]: typo: its

File: book/src/example/clone/csi-clone.yaml [hunk 8]
```diff
@@ -0,0 +1,15 @@
+apiVersion: v1
```
Comment (msau42) [inline book/src/example/clone/csi-clone.yaml]: are these examples referenced anywhere?

File: book/src/external-provisioner.md (hunk 1 above)
Comment (j-griffith) [reply book/src/external-provisioner.md]: Yes, I'll look at doing that and adding sub-sections for the specific DS types.

**Comment by j-griffith** [comment]
Created a DataSource section under features, and placed snapshots and cloning under that subsection.

File: book/src/volume-cloning.md [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+## Deploying volume clone functionality
+
+The Kubernetes CSI development team maintains the [external-provisioner](external-provisioner.md) which is responsible for detecting requests for a PVC DataSource and providing that information to the plugin during provisioning in the call to the plugins `CreateVolume` function.  It's up to the plugin to check the `csi.CreateVolumeRequest` for a `VolumeContentSource` entry in the CreateVolumeRequest object.
+
+There are no additional side-cars or add on components required.
+
+## Enabling Volume expansion for CSI volumes in Kubernetes
+
+Volume cloning for CSI volumes is an alpha feature (Kubernetes 1.15) and hence must be explicitly enabled via feature gate:
+This is spelled incorrectly.
```
Comment (msau42) [inline book/src/volume-cloning.md]: What is spelled incorrectly?

File: book/src/volume-datasources.md [hunk 10]
```diff
@@ -0,0 +1,15 @@
+## Kubernetes PVC DataSource (CSI VolumeContentSource)
+
+When creating a new PersistentVolumeClaim, the Kubernetes API provides a `PersistentVolumeClaim.DataSource` parameter.  This parameter is used to specify the CSI `CreateVolumeRequest.VolumeContentSource` option for CSI Provisioners The `VolumeContentSource` parameter instructs the CSI plugin to pre-populate the volume being provisioned with data from the specified source.
```
Comment (msau42) [inline book/src/volume-datasources.md]: period after Provisioners.

File: book/src/volume-datasources.md [hunk 11]
```diff
@@ -0,0 +1,15 @@
+## Kubernetes PVC DataSource (CSI VolumeContentSource)
+
+When creating a new PersistentVolumeClaim, the Kubernetes API provides a `PersistentVolumeClaim.DataSource` parameter.  This parameter is used to specify the CSI `CreateVolumeRequest.VolumeContentSource` option for CSI Provisioners The `VolumeContentSource` parameter instructs the CSI plugin to pre-populate the volume being provisioned with data from the specified source. 
+
+### External Provisioner Responsibilities
+
+If a `DataSource` is specified in the `CreateVolume` call to the CSI external provisioner, the external provisioner will evaluate whether the type of DataSource is supported, and if the data contained within the reference object is valid.  If the specified `DataSource.Kind` is supported, and the object is valid, the external provisioner will attempt to fetch the specified resource, ensuring that exists and obtaining the appropriate ID of the object to pass to the plugin.
```
Comment (msau42) [inline book/src/volume-datasources.md]: the last part can be simplified to "the external provisioner will fetch the resource and pass the appropriate object id to the plugin"

File: book/src/external-provisioner.md (hunk 1 above)
Comment (msau42) [reply book/src/external-provisioner.md]: In this page can you also add a `### DataSources` subsection to put these 2 paragraphs under?  And then point to the data sources page?

File: book/src/volume-cloning.md (hunk 9 above)
Comment (j-griffith) [reply book/src/volume-cloning.md]: :)

File: book/src/SUMMARY.md [hunk 12]
```diff
@@ -16,13 +16,15 @@
...
     - [Features](features.md)
         - [Secrets & Credentials](secrets-and-credentials.md)
-        - [Volume Snapshot & Restore](snapshot-restore-feature.md)
         - [Topology](topology.md)
         - [Raw Block Volume](raw-block.md)
         - [Skip Attach](skip-attach.md)
         - [Pod Info on Mount](pod-info.md)
         - [Ephemeral Local Volumes](ephemeral-local-volumes.md)
         - [Volume expansion](volume-expansion.md)
+        - [DataSources](volume-datasources.md)
```
Comment (msau42) [inline book/src/SUMMARY.md]: Add a space "Data Sources"

File: book/src/external-provisioner.md [hunk 13]
```diff
@@ -23,7 +23,13 @@ Once a new volume is successfully provisioned, the sidecar container creates a K
 
 The deletion of a `PersistentVolumeClaim` object bound to a `PersistentVolume` corresponding to this driver with a `delete` reclaim policy causes the sidecar container to trigger a `DeleteVolume` operation against the specified CSI endpoint to delete the volume. Once the volume is successfully deleted, the sidecar container also deletes the `PersistentVolume` object representing the volume.
 
-The CSI `external-provisioner` also supports the `Snapshot` DataSource. If a `Snapshot` CRD is specified as a data source on a PVC object, the sidecar container fetches the information about the snapshot by fetching the `SnapshotContent` object and populates the data source field in the resulting `CreateVolume` call to indicate to the storage system that the new volume should be populated using the specified snapshot.
+### DataSources
```
Comment (msau42) [inline book/src/external-provisioner.md]: Maybe start off with one sentence like "The external-provisioner can provision a volume that is pre-populated from a data source"

File: book/src/volume-cloning.md [hunk 14]
```diff
@@ -0,0 +1,48 @@
+# Volume Cloning
```
Comment (msau42) [inline book/src/volume-cloning.md]: The comments on this page still need to be addressed.

File: book/src/SUMMARY.md (hunk 12 above)
Comment (j-griffith) [reply book/src/SUMMARY.md]: done

File: book/src/example/clone/csi-clone.yaml (hunk 8 above)
Comment (j-griffith) [reply book/src/example/clone/csi-clone.yaml]: They were not, I'll add an Example header in the cloning doc and point to them there.

File: book/src/external-provisioner.md (hunk 13 above)
Comment (j-griffith) [reply book/src/external-provisioner.md]: I broke this up and hopefully it flows better.

File: book/src/external-provisioner.md (hunk 1 above)
Comment (j-griffith) [reply book/src/external-provisioner.md]: Yes, I have that actually; but it's not in this update.

File: book/src/volume-cloning.md (hunk 2 above)
Comment (j-griffith) [reply book/src/volume-cloning.md]: sorry, I missed that; added

File: book/src/volume-cloning.md (hunk 3 above)
Comment (j-griffith) [reply book/src/volume-cloning.md]: done

File: book/src/volume-cloning.md (hunk 4 above)
Comment (j-griffith) [reply book/src/volume-cloning.md]: Line #21 doesn't cut it?  I'll rework it.

File: book/src/volume-cloning.md (hunk 6 above)
Comment (j-griffith) [reply book/src/volume-cloning.md]: geesh

File: book/src/volume-cloning.md (hunk 7 above)
Comment (j-griffith) [reply book/src/volume-cloning.md]: changed

File: book/src/external-provisioner.md [hunk 15]
```diff
@@ -23,7 +23,20 @@ Once a new volume is successfully provisioned, the sidecar container creates a K
...
+
+#### Snapshot
+
+The CSI `external-provisioner` supports the `Snapshot` DataSource. If a `Snapshot` CRD is specified as a data source on a PVC object, the sidecar container fetches the information about the snapshot by fetching the `SnapshotContent` object and populates the data source field in the resulting `CreateVolume` call to indicate to the storage system that the new volume should be populated using the specified snapshot.
+
+#### PersistentVolumeClaim (clone)
+
+Cloning is also implemented by specifying a `kind:` of type `PersistentVolumeClaim` in the DataSource field of a Provision request.  It's the responsbility of the external-provisioner to verify that the claim specified in the DataSource object exists, is in the same storage class as the volume being provisioned and that the claim is currently `Bound`.
+
+For more information on how data sources are handled see [DataSources](volume-datasources.md).
```
Comment (msau42) [inline book/src/external-provisioner.md]: This sentence can be removed

File: book/src/volume-cloning.md [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+
+## Status and Releases
+
+** Status:** Alpha
+
+Compatible with CSI Version | Supported CSI external Provisioner Version | Min k8s Version | Max k8s version
+--|--|--|--|--|--
+v1.0  | v1.3 | v1.15 | -
```
Comment (msau42) [inline book/src/volume-cloning.md]: We've been leaving out the 'v' in the other pages

File: book/src/volume-cloning.md [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+control-plane CSI RPC call.
+
+## Implementing Volume cloning functionality
+
+To implement volume cloning the CSI driver MUST:
+
+1. Implement checks for `csi.CreateVolumeRequest.VolumeContentSource` in the plugins `CreateVolume` function implementation.
+2. Implement `CLONE_VOLUME` controller capability.
+
+The cloning feature in Kubernetes is implemented by adding a `DataSource` entry of `kind: PersistentVolumeClaim` to the `csi.CreateVolumeRequest`.  This presents the source volume (PersistentVolumeClaim to clone from) as a generic data source.  The source volume must:
```
Comment (msau42) [inline book/src/volume-cloning.md]: Is this information already captured in the external-provisioner and/or Kubernetes user docs?

This is more from the user's perspective, while this section is about how a vendor implements cloning in their driver.

File: book/src/volume-cloning.md [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+A Clone is defined as a duplicate of an existing Kubernetes Volume.  For more information on cloning in Kubernetes see the concepts doc for [Volume Cloning](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#volume-cloning).
+
+A storage provider that allows volume cloning as a create feature, may choose to implement volume cloning via a
+control-plane CSI RPC call.
+
+## Implementing Volume cloning functionality
+
+To implement volume cloning the CSI driver MUST:
+
+1. Implement checks for `csi.CreateVolumeRequest.VolumeContentSource` in the plugins `CreateVolume` function implementation.
```
Comment (msau42) [inline book/src/volume-cloning.md]: typo: plugin's

File: book/src/volume-cloning.md [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+
+```
+--feature-gates=VolumePVCDataSource=true
+```
+
+## Example implementation
+
+A trivial example implementation can be found in the [csi-hostpath plugin](https://github.com/kubernetes-csi/csi-driver-host-path) in its implementation of `CreateVolume`.
+
+## Examples
```
Comment (msau42) [inline book/src/volume-cloning.md]: Is this already included in Kubernetes user docs?

File: book/src/external-provisioner.md (hunk 15 above)
Comment (j-griffith) [reply book/src/external-provisioner.md]: removed

File: book/src/volume-cloning.md (hunk 16 above)
Comment (j-griffith) [reply book/src/volume-cloning.md]: Ok, the only page I saw using this at all (first glance any way) was the external-resizer page https://kubernetes-csi.github.io/docs/external-resizer.html

I'll remove them; this pr has me thinking we could use some doc writing guides and clean up quite a few inconsistencies.

File: book/src/volume-cloning.md (hunk 18 above)
Comment (j-griffith) [reply book/src/volume-cloning.md]: pluralization of a regular noun, but I make so many typos I'm the last person to argue

File: book/src/volume-cloning.md (hunk 17 above)
Comment (j-griffith) [reply book/src/volume-cloning.md]: It is in the k8s docs and the k8s blog (or will be when 1.15 releases and things update).  I'll remove it from here, the two sets of docs are sort of colliding at this point.

File: book/src/volume-cloning.md (hunk 19 above)
Comment (j-griffith) [reply book/src/volume-cloning.md]: yes, I'll remove line #56 as well as delete the example yaml files that I had added.

File: book/src/volume-cloning.md (hunk 16 above)
Comment (msau42) [reply book/src/volume-cloning.md]: bleh. All the "feature" pages leave out the 'v': https://kubernetes-csi.github.io/docs/topology.html

We'll have to do another sweep later and make them consistent.

File: book/src/volume-cloning.md [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+For details regarding the kuberentes API for volume cloning, please see [kubernetes concepts](https://kubernetes.io/docs/concepts/storage/volume-pvc-datasource/).
+
+## Implementing Volume cloning functionality
+
+To implement volume cloning the CSI driver MUST:
+
+1. Implement checks for `csi.CreateVolumeRequest.VolumeContentSource` in the plugin's `CreateVolume` function implementation.
+2. Implement `CLONE_VOLUME` controller capability.
+
+Note that the `external-provisioner` side car currently handles the checks listed above, as well as verifying the specified source can be fetched using the Kubernetes API.
```
Comment (msau42) [inline book/src/volume-cloning.md]: which checks above?

File: book/src/volume-cloning.md [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+
+It is the responsibility of the storage plugin to either implement an expansion after clone if a provision request size is greater than the source.
```
Comment (msau42) [inline book/src/volume-cloning.md]: Or can you let the external-resizer handle the resizing independently after clone?

File: book/src/volume-cloning.md (hunk 20 above)
Comment (j-griffith) [reply book/src/volume-cloning.md]: Yeah... the list of checks that I just removed and left up to the users docs in k8s

File: book/src/volume-cloning.md (hunk 21 above)
Comment (j-griffith) [reply book/src/volume-cloning.md]: arg... yes, I'll add that as a possible option.

File: book/src/SUMMARY.md [hunk 22]
```diff
@@ -16,13 +16,15 @@
...
-        - [Volume Snapshot & Restore](snapshot-restore-feature.md)
         - [Topology](topology.md)
         - [Raw Block Volume](raw-block.md)
         - [Skip Attach](skip-attach.md)
         - [Pod Info on Mount](pod-info.md)
         - [Ephemeral Local Volumes](ephemeral-local-volumes.md)
         - [Volume expansion](volume-expansion.md)
+        - [Data Sources](volume-datasources.md)
+	        - [Cloning](volume-cloning.md)
+            - [Volume Snapshot & Restore](snapshot-restore-feature.md)
```
Comment (msau42) [inline book/src/SUMMARY.md]: nit: indent is off

File: book/src/volume-cloning.md [hunk 23]
```diff
@@ -0,0 +1,48 @@
+# Volume Cloning
+
+## Status and Releases
+
+** Status:** Alpha
```
Comment (msau42) [inline book/src/volume-cloning.md]: Can you also put the status in the table, similar to: https://kubernetes-csi.github.io/docs/topology.html

File: book/src/SUMMARY.md (hunk 22 above)
Comment (j-griffith) [reply book/src/SUMMARY.md]: ahhh, I thought that was a git diff artifact, but some tabs snuck in on line 25, fixed.

File: book/src/volume-cloning.md (hunk 23 above)
Comment (j-griffith) [reply book/src/volume-cloning.md]: Yep, I'll actually match how you have the table in topology

Comment (msau42) [reply book/src/volume-cloning.md]: And this line can be removed now :-)

File: book/src/volume-cloning.md (hunk 19 above)
Comment (msau42) [reply book/src/volume-cloning.md]: still needs to be removed?

File: book/src/volume-cloning.md [hunk 24]
```diff
@@ -0,0 +1,46 @@
...
+For details regarding the kuberentes API for volume cloning, please see [kubernetes concepts](https://kubernetes.io/docs/concepts/storage/volume-pvc-datasource/).
+
+## Implementing Volume cloning functionality
+
+To implement volume cloning the CSI driver MUST:
+
+1. Implement checks for `csi.CreateVolumeRequest.VolumeContentSource` in the plugin's `CreateVolume` function implementation.
+2. Implement `CLONE_VOLUME` controller capability.
+
+It is the responsibility of the storage plugin to either implement an expansion after clone if a provision request size is greater than the source, or to allow the external-resizer handle it.  In the case that the plugin does not support resize capability and it does not have the capability to create a clone that is great in size than the specified source volume, then the provision request should result in a failure.
```
Comment (msau42) [inline book/src/volume-cloning.md]: typo: greater

File: book/src/volume-cloning.md (hunk 23 above)
Comment (j-griffith) [reply book/src/volume-cloning.md]: DeRp!

File: book/src/volume-cloning.md (hunk 19 above)
Comment (j-griffith) [reply book/src/volume-cloning.md]: For real this time!

File: book/src/volume-cloning.md (hunk 24 above)
Comment (j-griffith) [reply book/src/volume-cloning.md]: ?

Comment (msau42) [reply book/src/volume-cloning.md]: great => greater

Comment (j-griffith) [reply book/src/volume-cloning.md]: Ah, sorry I was stuck at "expansion after clone if a provision request size is greater than the"

**Pull request merged**